File: backend/service/referenceService/referenceService.go
package referenceService

import (
	"errors"
	"permit-app/model"
	"permit-app/repo/referenceCategoryRepository"
	"permit-app/repo/referenceRepository"

	"gorm.io/gorm"
)

type ReferenceService interface {
	CreateReference(req model.ReferenceRequest) (*model.ReferenceResponse, error)
	GetReferences(req model.ReferenceListRequest) ([]model.ReferenceResponse, int64, error)
	GetReferenceByID(id int64) (*model.ReferenceResponse, error)
	GetReferencesByCategoryID(categoryID int64) ([]model.ReferenceResponse, error)
	GetReferencesByModuleID(moduleID int64) ([]model.ReferenceResponse, error)
	UpdateReference(id int64, req model.ReferenceRequest) (*model.ReferenceResponse, error)
	DeleteReference(id int64) error
}

type referenceServiceImpl struct {
	referenceRepo referenceRepository.ReferenceRepository
	categoryRepo  referenceCategoryRepository.ReferenceCategoryRepository
}

func NewReferenceService(
	referenceRepo referenceRepository.ReferenceRepository,
	categoryRepo referenceCategoryRepository.ReferenceCategoryRepository,
) ReferenceService {
	return &referenceServiceImpl{
		referenceRepo: referenceRepo,
		categoryRepo:  categoryRepo,
	}
}

func (s *referenceServiceImpl) CreateReference(req model.ReferenceRequest) (*model.ReferenceResponse, error) {
	// Validate category exists
	_, err := s.categoryRepo.FindByID(req.ReferenceCategoryID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("reference category not found")
		}
		return nil, err
	}

	reference := &model.Reference{
		ReferenceCategoryID: req.ReferenceCategoryID,
		Name:                req.Name,
		IsActive:            true,
	}

	if req.IsActive != nil {
		reference.IsActive = *req.IsActive
	}

	if err := s.referenceRepo.Create(reference); err != nil {
		return nil, err
	}

	// Reload with relations
	reference, err = s.referenceRepo.FindByID(reference.ID)
	if err != nil {
		return nil, err
	}

	response := model.ToReferenceResponse(reference)
	return &response, nil
}

func (s *referenceServiceImpl) GetReferences(req model.ReferenceListRequest) ([]model.ReferenceResponse, int64, error) {
	filters := make(map[string]interface{})

	if req.ReferenceCategoryID > 0 {
		filters["reference_category_id"] = req.ReferenceCategoryID
	}
	if req.ModuleID > 0 {
		filters["module_id"] = req.ModuleID
	}
	if req.Name != "" {
		filters["name"] = req.Name
	}
	if req.IsActive != nil {
		filters["is_active"] = *req.IsActive
	}

	references, total, err := s.referenceRepo.FindAll(filters, req.Page, req.Limit)
	if err != nil {
		return nil, 0, err
	}

	// Convert to response
	responses := make([]model.ReferenceResponse, len(references))
	for i, reference := range references {
		responses[i] = model.ToReferenceResponse(&reference)
	}

	return responses, total, nil
}

func (s *referenceServiceImpl) GetReferenceByID(id int64) (*model.ReferenceResponse, error) {
	reference, err := s.referenceRepo.FindByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("reference not found")
		}
		return nil, err
	}

	response := model.ToReferenceResponse(reference)
	return &response, nil
}

func (s *referenceServiceImpl) GetReferencesByCategoryID(categoryID int64) ([]model.ReferenceResponse, error) {
	references, err := s.referenceRepo.FindByCategoryID(categoryID)
	if err != nil {
		return nil, err
	}

	responses := make([]model.ReferenceResponse, len(references))
	for i, reference := range references {
		responses[i] = model.ToReferenceResponse(&reference)
	}

	return responses, nil
}

func (s *referenceServiceImpl) GetReferencesByModuleID(moduleID int64) ([]model.ReferenceResponse, error) {
	references, err := s.referenceRepo.FindByModuleID(moduleID)
	if err != nil {
		return nil, err
	}

	responses := make([]model.ReferenceResponse, len(references))
	for i, reference := range references {
		responses[i] = model.ToReferenceResponse(&reference)
	}

	return responses, nil
}

func (s *referenceServiceImpl) UpdateReference(id int64, req model.ReferenceRequest) (*model.ReferenceResponse, error) {
	reference, err := s.referenceRepo.FindByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("reference not found")
		}
		return nil, err
	}

	// Validate category exists if changed
	if req.ReferenceCategoryID != reference.ReferenceCategoryID {
		_, err := s.categoryRepo.FindByID(req.ReferenceCategoryID)
		if err != nil {
			if errors.Is(err, gorm.ErrRecordNotFound) {
				return nil, errors.New("reference category not found")
			}
			return nil, err
		}
	}

	reference.ReferenceCategoryID = req.ReferenceCategoryID
	reference.Name = req.Name
	if req.IsActive != nil {
		reference.IsActive = *req.IsActive
	}

	if err := s.referenceRepo.Update(reference); err != nil {
		return nil, err
	}

	// Reload with relations
	reference, err = s.referenceRepo.FindByID(reference.ID)
	if err != nil {
		return nil, err
	}

	response := model.ToReferenceResponse(reference)
	return &response, nil
}

func (s *referenceServiceImpl) DeleteReference(id int64) error {
	reference, err := s.referenceRepo.FindByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("reference not found")
		}
		return err
	}

	return s.referenceRepo.Delete(reference.ID)
}

File: frontend/src/services/reference.service.ts
import { apiClient } from "@/lib/api-client";
import { ApiResponse } from "@/types/api";
import {
  Reference,
  ReferenceRequest,
  ReferenceListRequest,
} from "@/types/reference";

export const referenceService = {
  async getAll(
    params: ReferenceListRequest = {}
  ): Promise<
    ApiResponse<Reference[]> & {
      meta?: { page: number; limit: number; total: number };
    }
  > {
    const queryParams = new URLSearchParams();

    if (params.reference_category_id)
      queryParams.append(
        "reference_category_id",
        params.reference_category_id.toString()
      );
    if (params.module_id)
      queryParams.append("module_id", params.module_id.toString());
    if (params.name) queryParams.append("name", params.name);
    if (params.is_active !== undefined)
      queryParams.append("is_active", params.is_active.toString());
    if (params.page) queryParams.append("page", params.page.toString());
    if (params.limit) queryParams.append("limit", params.limit.toString());

    const response = await apiClient.get<ApiResponse<Reference[]>>(
      `/references?${queryParams.toString()}`
    );
    return response as any;
  },

  async getById(id: number): Promise<Reference> {
    const response = await apiClient.get<ApiResponse<Reference>>(
      `/references/${id}`
    );
    return response.data;
  },

  async getByCategoryId(categoryId: number): Promise<Reference[]> {
    const response = await apiClient.get<ApiResponse<Reference[]>>(
      `/reference-categories/${categoryId}/references`
    );
    return response.data;
  },

  async getByModuleId(moduleId: number): Promise<Reference[]> {
    const response = await apiClient.get<ApiResponse<Reference[]>>(
      `/modules/${moduleId}/references`
    );
    return response.data;
  },

  async create(data: ReferenceRequest): Promise<Reference> {
    const response = await apiClient.post<ApiResponse<Reference>>(
      "/references",
      data
    );
    return response.data;
  },

  async update(id: number, data: ReferenceRequest): Promise<Reference> {
    const response = await apiClient.put<ApiResponse<Reference>>(
      `/references/${id}`,
      data
    );
    return response.data;
  },

  async delete(id: number): Promise<void> {
    await apiClient.delete(`/references/${id}`);
  },
};

